scylla/src/transport/load_balancing/default.rs

```diff
@@ -393,16 +423,24 @@ impl DefaultPolicyBuilder {
     /// Builds a new DefaultPolicy with the previously set configuration.
     pub fn build(self) -> Arc<dyn LoadBalancingPolicy> {
         let latency_awareness = self.latency_awareness.map(|builder| builder.build());
+        let preferred_rack = self.preferred_rack.clone();
         let pick_predicate = if let Some(ref latency_awareness) = latency_awareness {
             let latency_predicate = latency_awareness.generate_predicate();
-            Box::new(move |node: &NodeRef| DefaultPolicy::is_alive(node) && latency_predicate(node))
-                as Box<dyn Fn(&NodeRef) -> bool + Send + Sync + 'static>
+            Box::new(move |node: &NodeRef| {
+                DefaultPolicy::is_alive(node)
+                    && latency_predicate(node)
+                    && DefaultPolicy::rack_predicate(node, &preferred_rack)
+            }) as Box<dyn Fn(&NodeRef) -> bool + Send + Sync + 'static>
         } else {
-            Box::new(DefaultPolicy::is_alive)
+            Box::new(move |node: &NodeRef<'_>| {
+                DefaultPolicy::is_alive(node)
+                    && DefaultPolicy::rack_predicate(node, &preferred_rack)
+            })
```

Comment:
I'm afraid this defies the whole pick mechanism. Please note that `rack_predicate()` evaluates to false for `preferred_rack` set to `None`. Having this in mind, we can see that if no rack is preferred, then `pick_predicate` is always false, so the `pick()` function will be able to choose a node only not using the `pick_predicate()`, in line with comment: 
```
// Previous checks imply that every node we could have selected is down.
// Let's try to return a down node that wasn't disabled.
```
...which is surely not the desired behaviour.

Hence, either `pick_predicate` must not call `rack_predicate` or, even better, `rack_predicate` shall be true if no rack is preferred.